ibis/expr/datatypes/cast.py

```diff
@@ -145,3 +145,26 @@ def highest_precedence(dtypes: Iterator[dt.DataType]) -> dt.DataType:
         return functools.reduce(higher_precedence, collected)
     else:
         return dt.null
+
+
+@public
+def same_base_type(source: dt.DataType, target: dt.DataType):
+    if (
+        (source.is_boolean() and target.is_boolean())
+        or (source.is_integer() and target.is_integer())
+        or (source.is_floating() and target.is_floating())
+        or (source.is_decimal() and target.is_decimal())
+        or (source.is_string() and target.is_string())
+        or (source.is_uuid() and target.is_uuid())
+        or (source.is_date() and target.is_date())
+        or (source.is_timestamp() and target.is_timestamp())
+        or (source.is_interval() and target.is_interval())
+        or (source.is_time() and target.is_time())
+        or (source.is_json() and target.is_json())
+        or (source.is_array() and target.is_array())
+        or (source.is_map() and target.is_map())
+        or (source.is_struct() and target.is_struct())
+        or (source.is_geospatial() and target.is_geospatial())
+    ):
+        return True
+    return False
```

Comment:
```suggestion
    return type(source) == type(target)
```

Would that work?

Reply:
I think in the case of 1000 and 0 one would return int16 and the other would return int8, so we'd have to check whether both inherit from the same immediate parent class, which is also an alternative but I'm wondering if this has more complexity involved

Reply:
Ah right.

Reply:
The alternative would be to check the builtin type of the underlying value (instead of Ibis type), but again I think this might just be simpler 😅

Reply:
The issue with this implementation is that it adds a place outside of the core datatypes module where we have to remember to update the code if we ever add a datatype. Given how infrequently we add a type it's unlikely we'll remember to update this code.

I think an abstract method on the `DataType` base class method `is_same_base_type(self, other)` that has the individual cases will help avoid creating data types that do not implement this functionality.

It'll be more code, but easier to maintain IMO.